flow/project.py

```diff
@@ -2636,34 +2942,51 @@ def select(operation):
             if not operations:
                 break   # No more pending operations or execution limits reached.
 
+            def key_func_by_job(op):
+                # In order to group the aggregates in a by-job manner, we need
+                # to first sort the aggregates using their aggregate id.
+                return get_aggregate_id(op._jobs)
+
             # Optionally re-order operations for execution if order argument is provided:
             if callable(order):
                 operations = list(sorted(operations, key=order))
             elif order == 'cyclic':
                 groups = [list(group)
-                          for _, group in groupby(operations, key=lambda op: op._jobs)]
+                          for _, group in groupby(sorted(operations, key=key_func_by_job),
+                                                  key=key_func_by_job)]
                 operations = list(roundrobin(*groups))
             elif order == 'random':
                 random.shuffle(operations)
-            elif order is None or order in ('none', 'by-job'):
-                pass  # by-job is the default order
+            elif order == 'by-job':
+                groups = [list(group)
+                          for _, group in groupby(sorted(operations, key=key_func_by_job),
+                                                  key=key_func_by_job)]
```

Comment:
this part is the same for `cyclic` besides the `roundrobin`. it may be worth combining into one branch of the conditional with an extra nesting level.